### PR TITLE
flesh out super() handling a bit more

### PR DESCRIPTION
**Pull request by toshok**
super() can be partially applied, and it acts as a descriptor, filling in the
missing pieces.

test/tests/super_getattribute.py is adapted from part of the cpython test suite

### CODE REVIEW

File: src/runtime/super.cpp [hunk 1]
```diff
@@ -128,10 +128,10 @@ Box* superGetattribute(Box* _s, Box* _attr) {
...
     }
 
-    Box* r = typeLookup(s->cls, std::string(attr->s()), NULL);
-    // TODO implement this
-    RELEASE_ASSERT(r, "should call the equivalent of objectGetattr here");
-    return processDescriptor(r, s, s->cls);
+    Box* r = getattrInternalGeneric(s, attr->s(), NULL, true, false, NULL, NULL);
+    if (r)
+        return processDescriptor(r, s, s->cls);
+    raiseExcHelper(AttributeError, "'super' object has no attribute '%s'", attr->data());
```
Comment (kmod) [inline src/runtime/super.cpp]: Can we do this the way CPython does?

File: src/runtime/super.cpp [hunk 2]
```diff
@@ -187,6 +187,36 @@ Box* superInit(Box* _self, Box* _type, Box* obj) {
...
+           call its type */
+        return runtimeCall(su->cls, ArgPassSpec(2, 0, false, false), su->type, obj, NULL, NULL, NULL);
+    } else {
+        /* Inline the common case */
+        BoxedClass* obj_type = supercheck(su->type, obj);
+        if (obj_type == NULL)
+            return NULL;
+        newobj = new (su->cls) BoxedSuper(NULL, NULL, NULL);
+        if (newobj == NULL)
+            return NULL;
```
Comment (kmod) [inline src/runtime/super.cpp]: These should be `throwCapiException()`.  I usually prefer though to just leave the CPython code unmodified, ie have this be a capi-style function and stick it into tp_descr_get.

**Comment by toshok** [comment]
hm, sorry about that - I could have sworn I added a comment to this PR :/

to paraphrase - tp_descr_get inheritance isn't working for us.  subclasses of super() get tp_descr_get set to NULL.  I was planning to get the feature done using `__get__` then go back and fix tp_descr_get to work, and update things here.

Given that there were problems with the code partially ported to `__get__`, I'll go back and fix tp_descr_get inheritance and update this PR.

**Comment by kmod** [comment]
We should call `add_operators(super_cls)` right before `super_cls->freeze()` -- I think that might be the source of the tp_ inheritance issue?  Sorry for the landmine; I think we've gotten to the point that freeze() could directly call add_operators, but I'm not sure.

**Pull request closed without merging**